### DISCUSSION TITLE
交叉编译Android的Tquic库

### DISCUSSION THREAD

**Discussion by stopandjoin** [Tencent/tquic#190] [Q&A (问题和回答)] [answered]
请教一下大家，https://tquic.net/docs/getting_started/installation  这个文档下描述的Android编译的方法，是在windows 系统下编译的，还是linux系统下编译的？  我在windows 系统下按照指引，安装了rust环境，VS环境， 然后在命令行窗口编译会报错：
![image](https://github.com/Tencent/tquic/assets/164285863/dd6a5e13-ed60-467b-a961-0941731b58b4)

**Comment by glime-ai** [comment]
根据提供的信息和文档，https://tquic.net/docs/getting_started/installation 这个文档主要描述了在Linux、MacOS、FreeBSD、iOS和Android平台上编译TQUIC的方法，并没有直接提及Windows系统。TQUIC是基于Rust语言开发的，理论上它支持Rust支持的任何平台，但是官方文档中并没有特别指出对Windows的支持。

对于Android平台的编译，文档中提到了需要安装Rust工具链和cargo-ndk来简化Android库的构建过程。这些步骤通常在Linux或MacOS系统中进行，因为这些系统与Rust和Android NDK的兼容性较好。

如果你在Windows系统下遇到编译问题，可能是因为某些依赖或工具链在Windows上的配置与Linux或MacOS不同。虽然Rust和Android NDK都可以在Windows上使用，但可能需要额外的配置步骤或解决特定的兼容性问题。

建议查看官方GitHub仓库的issue或讨论区，看看是否有其他用户遇到类似问题以及如何解决。例如，可以查看是否有关于在Windows上编译TQUIC的讨论或指南。

此外，考虑使用Linux虚拟机或WSL（Windows Subsystem for Linux）在Windows上提供一个更兼容的环境进行编译。

最后，由于你提供的图片链接无法直接访问，无法直接查看具体的错误信息。如果问题仍然存在，建议将错误信息文本形式提供，以便更具体地分析问题。

**Comment by stopandjoin** [comment]
error: failed to run custom build command for `tquic v0.8.1 (E:\code\tquic-example-c\deps\tquic)`

Caused by:
  process didn't exit successfully: `E:\code\tquic-example-c\deps\tquic\target\release\build\tquic-5f5cedcd5192dd35\build-script-build` (exit code: 101)
  --- stdout
  CMAKE_GENERATOR_aarch64-linux-android = None
  CMAKE_GENERATOR_aarch64_linux_android = None
  TARGET_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  Info: default compiler flags are disabled
  Info: default compiler flags are disabled
  Info: default compiler flags are disabled
  CMAKE_PREFIX_PATH_aarch64-linux-android = None
  CMAKE_PREFIX_PATH_aarch64_linux_android = None
  TARGET_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_aarch64-linux-android = None
  CMAKE_aarch64_linux_android = None
  TARGET_CMAKE = None
  CMAKE = None
  running: "cmake" "E:\\code\\tquic-example-c\\deps\\tquic\\deps/boringssl" "-DANDROID_ABI=arm64-v8a" "-DCMAKE_TOOLCHAIN_FILE=C:\\Users\\GW00346215\\AppData\\Local\\Android\\Sdk\\ndk\\26.0.10792818\\build/cmake/android.toolchain.cmake" "-DANDROID_NATIVE_API_LEVEL=26" "-DANDROID_STL=c++_shared" "-DCMAKE_INSTALL_PREFIX=E:\\code\\tquic-example-c\\deps\\tquic\\target\\aarch64-linux-android\\release\\build\\tquic-c2f718fdbbd7a936\\out" "-DCMAKE_C_FLAGS= --target=aarch64-linux-android26" "-DCMAKE_CXX_FLAGS= --target=aarch64-linux-android26" "-DCMAKE_ASM_FLAGS= --target=aarch64-linux-android26" "-DCMAKE_BUILD_TYPE=Release"
  -- Building for: Visual Studio 17 2022
  -- Configuring incomplete, errors occurred!

  --- stderr
  CMake Error at CMakeLists.txt:19 (project):
    Failed to run MSBuild command:

      C:/Program Files/Microsoft Visual Studio/2022/Community/MSBuild/Current/Bin/amd64/MSBuild.exe

    to get the value of VCTargetsPath:

      适用于 .NET Framework MSBuild 版本 17.9.8+b34f75857
      生成启动时间为 2024/3/22 11:30:48。

      节点 1 上的项目“E:\code\tquic-example-c\deps\tquic\target\aarch64-linux-android\release\build\tquic-c2f718fdbbd7a936\out\build\CMakeFiles\3.28.0-msvc1\VCTargetsPath.vcxproj”(默认目标)。
      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(839,5): error : The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project. [E:\code\tquic-example-c\deps\tquic\target\aarch64-linux-android\release\build\tquic-c2f718fdbbd7a936\out\build\CMakeFiles\3.28.0-msvc1\VCTargetsPath.vcxproj]
      已完成生成项目“E:\code\tquic-example-c\deps\tquic\target\aarch64-linux-android\release\build\tquic-c2f718fdbbd7a936\out\build\CMakeFiles\3.28.0-msvc1\VCTargetsPath.vcxproj”(默认目标)的操作 - 失败。

      生成失败。

      “E:\code\tquic-example-c\deps\tquic\target\aarch64-linux-android\release\build\tquic-c2f718fdbbd7a936\out\build\CMakeFiles\3.28.0-msvc1\VCTargetsPath.vcxproj”(默认目标) (1) ->
      (_CheckForInvalidConfigurationAndPlatform 目标) ->
        C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(839,5): error : The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project. [E:\code\tquic-example-c\deps\tquic\target\aarch64-linux-android\release\build\tquic-c2f718fdbbd7a936\out\build\CMakeFiles\3.28.0-msvc1\VCTargetsPath.vcxproj]

          0 个警告
          1 个错误

      已用时间 00:00:00.25


    Exit code: 1



  thread 'main' panicked at C:\Users\GW00346215\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cmake-0.1.50\src\lib.rs:1098:5:

  command did not execute successfully, got: exit code: 1

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...

**Comment by iyangsj** [comment, marked as answer]
@stopandjoin 截图提示cmake可能没有安装

建议在MacOS或Linux平台编译Andorid移动端库。
Windows平台目前暂未正式支持。

**Reply by stopandjoin** [reply]
好的，明白了，谢谢。
